ringbuf/reader.go

```diff
@@ -0,0 +1,255 @@
+	epollFd, err := unix.EpollCreate1(unix.EPOLL_CLOEXEC)
+	if err != nil {
+		return nil, fmt.Errorf("can't create epoll fd: %v", err)
+	}
+
+	var (
+		fds        = []int{epollFd}
+		maxEntries = int(array.MaxEntries())
+		ring       *ringbufEventRing
+	)
+
+	if maxEntries != 0 && (maxEntries&(maxEntries-1)) != 0 {
+		return nil, fmt.Errorf("Ringbuffer map size %d is not a power of two", maxEntries)
+	}
+	defer func() {
+		if err != nil {
+			// close epollFd and closeFd
+			for _, fd := range fds {
+				unix.Close(fd)
+			}
+			if ring != nil {
+				ring.Close()
+			}
+		}
+	}()
+
+	ring, err = newRingBufEventRing(array.FD(), maxEntries)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ringbuf ring: %v", err)
+	}
+
+	if err := addToEpoll(epollFd, ring.fd); err != nil {
+		return nil, err
+	}
+
+	closeFd, err := unix.Eventfd(0, unix.O_CLOEXEC|unix.O_NONBLOCK)
+	if err != nil {
+		return nil, err
+	}
+	fds = append(fds, closeFd)
+
+	if err := addToEpoll(epollFd, closeFd); err != nil {
+		return nil, err
+	}
+
+	pr := &Reader{
+		ring:    ring,
+		array:   array,
+		mapFd:   array.FD(),
+		epollFd: epollFd,
+		// Allocate extra event for closeFd
+		epollEvents: make([]unix.EpollEvent, 2),
+		pollTimeout: opts.pollTimeout,
+		closeFd:     closeFd,
+	}
+	runtime.SetFinalizer(pr, (*Reader).Close)
+	return pr, nil
+}
+
+// Close frees resources used by the reader.
+//
+// It interrupts calls to Read.
+func (pr *Reader) Close() error {
+	var err error
+	pr.closeOnce.Do(func() {
+		runtime.SetFinalizer(pr, nil)
+
+		// Interrupt Read() via the closeFd event fd.
+		var value [8]byte
+		internal.NativeEndian.PutUint64(value[:], 1)
+		_, err = unix.Write(pr.closeFd, value[:])
+		if err != nil {
+			err = fmt.Errorf("can't write event fd: %v", err)
+			return
+		}
+
+		// Acquire the lock. This ensures that Read isn't running.
+		pr.mu.Lock()
+		defer pr.mu.Unlock()
+
+		unix.Close(pr.epollFd)
+		unix.Close(pr.closeFd)
+		pr.epollFd, pr.closeFd = -1, -1
+
+		if pr.ring != nil {
+			pr.ring.Close()
+		}
+		pr.ring = nil
+
+		pr.array.Close()
+	})
+	if err != nil {
+		return fmt.Errorf("close RingbufReader: %w", err)
+	}
+	return nil
+}
+
+// Read the next record from the BPF ringbuf.
+//
+// Calling Close interrupts the function.
+func (pr *Reader) Read() (Record, error) {
+	pr.mu.Lock()
+	defer pr.mu.Unlock()
+
+	if pr.epollFd == -1 {
+		return Record{}, errClosed
+	}
+
+	for {
+		nEvents, err := unix.EpollWait(pr.epollFd, pr.epollEvents, pr.pollTimeout)
+		if temp, ok := err.(temporaryError); ok && temp.Temporary() {
+			// Retry the syscall if we we're interrupted, see https://github.com/golang/go/issues/20400
+			continue
+		}
+
+		if err != nil {
+			return Record{}, err
+		}
+
+		for _, event := range pr.epollEvents[:nEvents] {
+			if int(event.Fd) == pr.closeFd {
+				return Record{}, errClosed
+			}
+		}
+
+		pr.ring.loadConsumer()
+
+		record, err := readRecordFromRing(pr.ring)
+		if err == errBusy {
```

Comment:
`errors.Is`, same for other error comparisons.

Reply:
fixed